### ISSUE TITLE
eta fails to build in nixpkgs

### ISSUE THREAD

**Issue by bbarker** [typelead/eta#736]
Seems to be due to a missing dependency - not sure why.

### System set up

 - system: `"x86_64-linux"`
 - host os: `Linux 4.13.0-38-generic, Ubuntu, 16.04.3 LTS (Xenial Xerus)`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 1.11.16`
 - channels(brandon): `"nixpkgs-18.09pre137305.d91caac6c3e, nixos-18.09pre134828.62dc989963c"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/brandon/channels/nixpkgs`


### Build Error

Using [these instructions](https://eta-lang.org/docs/eta-concepts/getting-started/install-eta#nix-environment):

```
$ git clone --branch stable --recursive --depth 1 https://github.com/typelead/eta
$ cd eta
$ nix-shell -A eta-build-shell
```

End of build:

```
...
Warning: Hackage.Security.Client.Formats: could not find link destinations for:                                                                                                                                                                                                                                                                                
    :-                                                                                                                                                                                                                                                                                                                                                         
Warning: Hackage.Security.Util.Some: could not find link destinations for:                                                                                                                                                                                                                                                                                     
    Typed TypeOf                                                                                                                                                                                                                                                                                                                                               
Warning: Hackage.Security.Client.Verify: could not find link destinations for:                                                                                                                                                                                                                                                                                 
    Finaliser                                                                                                                                                                                                                                                                                                                                                  
Warning: Hackage.Security.Key.Env: could not find link destinations for:                                                                                                                                                                                                                                                                                       
    checkKeyEnvInvariant                                                                                                                                                                                                                                                                                                                                       
Warning: Hackage.Security.Client.Repository: could not find link destinations for:                                                                                                                                                                                                                                                                             
    :-                                                                                                                                                                                                                                                                                                                                                         
Warning: Hackage.Security.Client: could not find link destinations for:                                                                                                                                                                                                                                                                                        
    Pattern Replacement                                                                                                                                                                                                                                                                                                                                        
Warning: Hackage.Security.Client.Repository.Remote: could not find link destinations for:                                                                                                                                                                                                                                                                      
    withMirror getRemote multipleExitPoints                                                                                                                                                                                                                                                                                                                    
Warning: Hackage.Security.Server: could not find link destinations for:                                                                                                                                                                                                                                                                                        
    Pattern Replacement                                                                                                                                                                                                                                                                                                                                        
Documentation created: dist/doc/html/hackage-security/index.html,                                                                                                                                                                                                                                                                                              
dist/doc/html/hackage-security/hackage-security.txt                                                                                                                                                                                                                                                                                                            
installing                                                                                                                                                                                                                                                                                                                                                     
Installing library in                                                                                                                                                                                                                                                                                                                                          
/nix/store/dxan6x094ndi48852i8hg0z98n7k917l-hackage-security-0.5.2.2/lib/ghc-7.10.3/hackage-security-0.5.2.2                                                                                                                                                                                                                                                   
Creating package registration file:                                                                                                                                                                                                                                                                                                                            
/nix/store/dxan6x094ndi48852i8hg0z98n7k917l-hackage-security-0.5.2.2/lib/ghc-7.10.3/package.conf.d/hackage-security-0.5.2.2.conf                                                                                                                                                                                                                               
post-installation fixup                                                                                                                                                                                                                                                                                                                                        
shrinking RPATHs of ELF executables and libraries in /nix/store/dxan6x094ndi48852i8hg0z98n7k917l-hackage-security-0.5.2.2                                                                                                                                                                                                                                      
shrinking /nix/store/dxan6x094ndi48852i8hg0z98n7k917l-hackage-security-0.5.2.2/lib/ghc-7.10.3/hackage-security-0.5.2.2/HShackage-security-0.5.2.2-H6Rq1sjsdVb6C64KHMME1G.o                                                                                                                                                                                     
wrong ELF type                                                                                                                                                                                                                                                                                                                                                 
shrinking /nix/store/dxan6x094ndi48852i8hg0z98n7k917l-hackage-security-0.5.2.2/lib/ghc-7.10.3/hackage-security-0.5.2.2/libHShackage-security-0.5.2.2-H6Rq1sjsdVb6C64KHMME1G-gh                                                                                                                                                                                 
c7.10.3.so                                                                                                                                                                                                                                                                                                                                                     
strip is /nix/store/j75dgadrff2d1fyc4fczmcgqkid2imdx-binutils-2.30/bin/strip                                                                                                                                                                                                                                                                                   
stripping (with command strip and flags -S) in /nix/store/dxan6x094ndi48852i8hg0z98n7k917l-hackage-security-0.5.2.2/lib                                                                                                                                                                                                                                        
patching script interpreter paths in /nix/store/dxan6x094ndi48852i8hg0z98n7k917l-hackage-security-0.5.2.2                                                                                                                                                                                                                                                      
checking for references to /tmp/nix-build-hackage-security-0.5.2.2.drv-0 in /nix/store/dxan6x094ndi48852i8hg0z98n7k917l-hackage-security-0.5.2.2...                                                                                                                                                                                                            
wrong ELF type                                                                                                                                                                                                                                                                                                                                                 
shrinking RPATHs of ELF executables and libraries in /nix/store/iy79xhgpvd21np5500r078ndqyx1lgqf-hackage-security-0.5.2.2-doc                                                                                                                                                                                                                                  
strip is /nix/store/j75dgadrff2d1fyc4fczmcgqkid2imdx-binutils-2.30/bin/strip                                                                                                                                                                                                                                                                                   
patching script interpreter paths in /nix/store/iy79xhgpvd21np5500r078ndqyx1lgqf-hackage-security-0.5.2.2-doc                                                                                                                                                                                                                                                  
checking for references to /tmp/nix-build-hackage-security-0.5.2.2.drv-0 in /nix/store/iy79xhgpvd21np5500r078ndqyx1lgqf-hackage-security-0.5.2.2-doc...                                                                                                                                                                                                        
building path(s) ‘/nix/store/4s1zxcl9nd2w2cjfwlg4gjcj36s02k3j-shake-0.16.4-data’, ‘/nix/store/j2ipj8abbbzx79jbvvln8hcbvx0gc8dr-shake-0.16.4’, ‘/nix/store/v3rsbgglj4nc3hc59s54                                                                                                                                                                                 
npf8xym6ai40-shake-0.16.4-doc’                                                                                                                                                                                                                                                                                                                                 
setupCompilerEnvironmentPhase                                                                                                                                                                                                                                                                                                                                  
Build with /nix/store/m67v6l20w9nr7zk8ja4dv6az4bnyvi67-ghc-7.10.3.                                                                                                                                                                                                                                                                                             
unpacking sources                                                                                                                                                                                                                                                                                                                                              
unpacking source archive /nix/store/5076v0gfs40plrfc30ikc7smia1i3if7-shake-0.16.4.tar.gz                                                                                                                                                                                                                                                                       
source root is shake-0.16.4                                                                                                                                                                                                                                                                                                                                    
setting SOURCE_DATE_EPOCH to timestamp 1522869954 of file shake-0.16.4/src/Test/Tup/root.cfg                                                                                                                                                                                                                                                                   
patching sources                                                                                                                                                                                                                                                                                                                                               
compileBuildDriverPhase                                                                                                                                                                                                                                                                                                                                        
setupCompileFlags: -package-db=/tmp/nix-build-shake-0.16.4.drv-0/package.conf.d -j8 -threaded                                                                                                                                                                                                                                                                  
[1 of 1] Compiling Main             ( Setup.hs, /tmp/nix-build-shake-0.16.4.drv-0/Main.o )                                                                                                                                                                                                                                                                     
Linking Setup ...                                                                                                                                                                                                                                                                                                                                              
configuring                                                                                                                                                                                                                                                                                                                                                    
configureFlags: --verbose --prefix=/nix/store/j2ipj8abbbzx79jbvvln8hcbvx0gc8dr-shake-0.16.4 --libdir=$prefix/lib/$compiler --libsubdir=$pkgid --datadir=/nix/store/4s1zxcl9nd2                                                                                                                                                                                 
w2cjfwlg4gjcj36s02k3j-shake-0.16.4-data/share/ghc-7.10.3 --docdir=/nix/store/v3rsbgglj4nc3hc59s54npf8xym6ai40-shake-0.16.4-doc/share/doc --with-gcc=gcc --package-db=/tmp/nix-                                                                                                                                                                                 
build-shake-0.16.4.drv-0/package.conf.d --enable-split-objs --enable-library-profiling --disable-executable-profiling --enable-shared --disable-coverage --disable-executable-                                                                                                                                                                                 
dynamic --disable-tests --enable-library-vanilla --enable-library-for-ghci --extra-lib-dirs=/nix/store/v4fkzi98745wi8psjdvfsp8nn6ygx9ww-ncurses-6.1/lib --extra-lib-dirs=/nix/                                                                                                                                                                                 
store/hwkkw10kb5w600c3dvv061gn1ld5fbja-gmp-6.1.2/lib --extra-lib-dirs=/nix/store/v4fkzi98745wi8psjdvfsp8nn6ygx9ww-ncurses-6.1/lib                                                                                                                                                                                                                              
Configuring shake-0.16.4...                                                                                                                                                                                                                                                                                                                                    
Setup: At least the following dependencies are missing:                                                                                                                                                                                                                                                                                                        
semigroups >=0.18                                                                                                                                                                                                                                                                                                                                              
builder for ‘/nix/store/rqzyzwwl2lfbgkxrjxj56200m8znamac-shake-0.16.4.drv’ failed with exit code 1                                                                                                                                                                                                                                                             
cannot build derivation ‘/nix/store/cv5d1s1h4qn46bqd3rfyr8sn7n4d4xgy-eta-build-0.0.0.drv’: 1 dependencies couldn't be built                                                                                                                                                                                                                                    
error: build of ‘/nix/store/cv5d1s1h4qn46bqd3rfyr8sn7n4d4xgy-eta-build-0.0.0.drv’ failed                                                                                                                                                                                                                                                                       
/home/brandon/.nix-profile/bin/nix-shell: failed to build all dependencies                           
```

**Comment by rahulmutt** [comment]
@puffnfresh It's failing while building `shake-0.16.4` so it probably has something to do with `eta-build.nix`. Any ideas?

**Comment by puffnfresh** [comment]
@bbarker I checked out both nixpkgs d91caac6c3e and 62dc989963c, they give shake-0.16.3 instead of 0.16.4 - do you know where that version of Shake is coming from on your system?

**Comment by puffnfresh** [comment]
I also recommend getting rid of Nix 1.11.16 and upgrading to 2.0

**Comment by bbarker** [comment]
@puffnfresh No, I don't think I installed it via any other means, unless stack relies on it, which doesn't seem to be the case: https://github.com/NixOS/nixpkgs/issues/39296
But I'll try to upgrade to 2.0 first.

**Comment by bbarker** [comment]
@puffnfresh makes sense, thanks: 137305.d91caac6c3e

**Comment by puffnfresh** [comment]
Perfect thank you, I can reproduce with that commit. I'll work on a fix.

**Comment by puffnfresh** [comment]
@bbarker can you try my branch here?

https://github.com/puffnfresh/ghcvm/tree/nix/fixes

**Comment by bbarker** [comment]
@puffnfresh using ee76d75705eb1e9b1b2f40ba48d85fd3f986c38c in the tree you listed, I had a very similar but not identical error as before:

```
No lhc found
No lhc-pkg found
No pkg-config found
Using strip version 2.30 found on system at:
/nix/store/j75dgadrff2d1fyc4fczmcgqkid2imdx-binutils-2.30/bin/strip
Using tar found on system at:
/nix/store/inn1myxkdnw11k8bzfs32bjm9w1w90jy-gnutar-1.30/bin/tar
No uhc found
building
Building tasty-smallcheck-0.8.1...
Preprocessing library tasty-smallcheck-0.8.1...
[1 of 1] Compiling Test.Tasty.SmallCheck ( Test/Tasty/SmallCheck.hs, dist/build/Test/Tasty/SmallCheck.o )
[1 of 1] Compiling Test.Tasty.SmallCheck ( Test/Tasty/SmallCheck.hs, dist/build/Test/Tasty/SmallCheck.p_o )
In-place registering tasty-smallcheck-0.8.1...
running tests
Package has no test suites.
haddockPhase
Running Haddock for tasty-smallcheck-0.8.1...
Running hscolour for tasty-smallcheck-0.8.1...
Preprocessing library tasty-smallcheck-0.8.1...
Preprocessing library tasty-smallcheck-0.8.1...
Warning: The documentation for the following packages are not installed. No
links will be generated to these packages: fail-4.9.0.0
Haddock coverage:
 100% (  4 /  4) in 'Test.Tasty.SmallCheck'
Documentation created: dist/doc/html/tasty-smallcheck/index.html,
dist/doc/html/tasty-smallcheck/tasty-smallcheck.txt
installing
Installing library in
/nix/store/siny4byn71qxfc6fyp110b7fvp47hm90-tasty-smallcheck-0.8.1/lib/ghc-7.10.3/tasty-smallcheck-0.8.1
Creating package registration file:
/nix/store/siny4byn71qxfc6fyp110b7fvp47hm90-tasty-smallcheck-0.8.1/lib/ghc-7.10.3/package.conf.d/tasty-smallcheck-0.8.1.conf
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/siny4byn71qxfc6fyp110b7fvp47hm90-tasty-smallcheck-0.8.1
shrinking /nix/store/siny4byn71qxfc6fyp110b7fvp47hm90-tasty-smallcheck-0.8.1/lib/ghc-7.10.3/tasty-smallcheck-0.8.1/HStasty-smallcheck-0.8.1-AzoO97WQFD1GOADOP5w9ve.o
wrong ELF type
shrinking /nix/store/siny4byn71qxfc6fyp110b7fvp47hm90-tasty-smallcheck-0.8.1/lib/ghc-7.10.3/tasty-smallcheck-0.8.1/libHStasty-smallcheck-0.8.1-AzoO97WQFD1GOADOP5w9ve-ghc7.10.
3.so
strip is /nix/store/j75dgadrff2d1fyc4fczmcgqkid2imdx-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/siny4byn71qxfc6fyp110b7fvp47hm90-tasty-smallcheck-0.8.1/lib
patching script interpreter paths in /nix/store/siny4byn71qxfc6fyp110b7fvp47hm90-tasty-smallcheck-0.8.1
checking for references to /tmp/nix-build-tasty-smallcheck-0.8.1.drv-0 in /nix/store/siny4byn71qxfc6fyp110b7fvp47hm90-tasty-smallcheck-0.8.1...
wrong ELF type
shrinking RPATHs of ELF executables and libraries in /nix/store/mcz7bxl2x5s9yx89k2spyhx03kfgqyba-tasty-smallcheck-0.8.1-doc
strip is /nix/store/j75dgadrff2d1fyc4fczmcgqkid2imdx-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/mcz7bxl2x5s9yx89k2spyhx03kfgqyba-tasty-smallcheck-0.8.1-doc
checking for references to /tmp/nix-build-tasty-smallcheck-0.8.1.drv-0 in /nix/store/mcz7bxl2x5s9yx89k2spyhx03kfgqyba-tasty-smallcheck-0.8.1-doc...
building path(s) ‘/nix/store/gfvihrjdhrcyg22i19gxg0km0ndmsvfh-eta-build-0.0.0’, ‘/nix/store/lpxvixbmnp3nix9q5j5g2s4c3h23pyfw-eta-build-0.0.0-doc’
setupCompilerEnvironmentPhase
Build with /nix/store/m67v6l20w9nr7zk8ja4dv6az4bnyvi67-ghc-7.10.3.
unpacking sources
unpacking source archive /nix/store/ilsa1wz29yk5iixjmacrrxhlpjq6cgx6-shake
source root is shake
patching sources
compileBuildDriverPhase
setupCompileFlags: -package-db=/tmp/nix-build-eta-build-0.0.0.drv-0/package.conf.d -j8 -threaded
[1 of 1] Compiling Main             ( Setup.hs, /tmp/nix-build-eta-build-0.0.0.drv-0/Main.o )
Linking Setup ...
configuring
configureFlags: --verbose --prefix=/nix/store/gfvihrjdhrcyg22i19gxg0km0ndmsvfh-eta-build-0.0.0 --libdir=$prefix/lib/$compiler --libsubdir=$pkgid --docdir=/nix/store/lpxvixbmn
p3nix9q5j5g2s4c3h23pyfw-eta-build-0.0.0-doc/share/doc --with-gcc=gcc --package-db=/tmp/nix-build-eta-build-0.0.0.drv-0/package.conf.d --ghc-option=-j8 --enable-split-objs --e
nable-library-profiling --disable-executable-profiling --enable-shared --disable-coverage --disable-executable-dynamic --enable-tests --enable-library-vanilla --enable-librar
y-for-ghci --extra-lib-dirs=/nix/store/v4fkzi98745wi8psjdvfsp8nn6ygx9ww-ncurses-6.1/lib --extra-lib-dirs=/nix/store/hwkkw10kb5w600c3dvv061gn1ld5fbja-gmp-6.1.2/lib --extra-lib
-dirs=/nix/store/v4fkzi98745wi8psjdvfsp8nn6ygx9ww-ncurses-6.1/lib
Configuring eta-build-0.0.0...
Dependency base >=4.7 && <5: using base-4.8.2.0
Dependency containers >=0.5: using containers-0.5.6.2
Dependency directory -any: using directory-1.2.2.0
Dependency etlas-cabal -any: using etlas-cabal-1.3.0.0
Dependency shake -any: using shake-0.16.4
Setup: The following installed packages are broken because other packages they
depend on are missing. These broken packages must be rebuilt before they can
be used.
package shake-0.16.4 is broken due to missing package
semigroups-0.18.4-92400039811c1e88b6606fca4b4c93ac
builder for ‘/nix/store/p6k58fjgfp62rwmiw3drbwk6y0szb9l8-eta-build-0.0.0.drv’ failed with exit code 1
error: build of ‘/nix/store/p6k58fjgfp62rwmiw3drbwk6y0szb9l8-eta-build-0.0.0.drv’ failed
/home/brandon/.nix-profile/bin/nix-shell: failed to build all dependencies
```

**Comment by bbarker** [comment]
My efforts to upgrade to 2.0.1 [hit a snag](https://github.com/NixOS/nix/issues/2121).

**Comment by bbarker** [comment]
@puffnfresh Managed to upgrade to nix 2.0.1 (which is exciting on its own), however, the problem seems to persist as in both cases above. For your branch:

```
bhpfq6z3y32sv4c6hn-etlas-1.3.0.0/bin
patching script interpreter paths in /nix/store/w9iraa2kpdf7p7bhpfq6z3y32sv4c6hn-etlas-1.3.0.0
checking for references to /tmp/nix-build-etlas-1.3.0.0.drv-0 in /nix/store/w9iraa2kpdf7p7bhpfq6z3y32sv4c6hn-etlas-1.3.0.0...
wrong ELF type
shrinking RPATHs of ELF executables and libraries in /nix/store/sbgbp4rrlrjdlxwi84sc1bl68q7wf5xl-etlas-1.3.0.0-doc
strip is /nix/store/j75dgadrff2d1fyc4fczmcgqkid2imdx-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/sbgbp4rrlrjdlxwi84sc1bl68q7wf5xl-etlas-1.3.0.0-doc
checking for references to /tmp/nix-build-etlas-1.3.0.0.drv-0 in /nix/store/sbgbp4rrlrjdlxwi84sc1bl68q7wf5xl-etlas-1.3.0.0-doc...
building '/nix/store/xvvandq9rzf0z3gw64jd2k9i0cc3rc14-eta-build-0.0.0.drv'...
setupCompilerEnvironmentPhase
Build with /nix/store/m67v6l20w9nr7zk8ja4dv6az4bnyvi67-ghc-7.10.3.
unpacking sources
unpacking source archive /nix/store/ilsa1wz29yk5iixjmacrrxhlpjq6cgx6-shake
source root is shake
patching sources
compileBuildDriverPhase
setupCompileFlags: -package-db=/tmp/nix-build-eta-build-0.0.0.drv-0/package.conf.d -j8 -threaded
[1 of 1] Compiling Main             ( Setup.hs, /tmp/nix-build-eta-build-0.0.0.drv-0/Main.o )
Linking Setup ...
configuring
configureFlags: --verbose --prefix=/nix/store/p0ss5xahznzkfb7g0a5w4j3di500wlvs-eta-build-0.0.0 --libdir=$prefix/lib/$compiler --libsubd
ir=$pkgid --docdir=/nix/store/ixc12m53m1zgl6944i1x2qvj0a1lv6v5-eta-build-0.0.0-doc/share/doc --with-gcc=gcc --package-db=/tmp/nix-build
-eta-build-0.0.0.drv-0/package.conf.d --ghc-option=-j8 --enable-split-objs --enable-library-profiling --disable-executable-profiling --
enable-shared --disable-coverage --disable-executable-dynamic --enable-tests --enable-library-vanilla --enable-library-for-ghci --extra
-lib-dirs=/nix/store/v4fkzi98745wi8psjdvfsp8nn6ygx9ww-ncurses-6.1/lib --extra-lib-dirs=/nix/store/hwkkw10kb5w600c3dvv061gn1ld5fbja-gmp-
6.1.2/lib --extra-lib-dirs=/nix/store/v4fkzi98745wi8psjdvfsp8nn6ygx9ww-ncurses-6.1/lib
Configuring eta-build-0.0.0...
Dependency base >=4.7 && <5: using base-4.8.2.0
Dependency containers >=0.5: using containers-0.5.6.2
Dependency directory -any: using directory-1.2.2.0
Dependency etlas-cabal -any: using etlas-cabal-1.3.0.0
Dependency shake -any: using shake-0.16.4
Setup: The following installed packages are broken because other packages they
depend on are missing. These broken packages must be rebuilt before they can
be used.
package shake-0.16.4 is broken due to missing package
semigroups-0.18.4-92400039811c1e88b6606fca4b4c93ac
builder for '/nix/store/xvvandq9rzf0z3gw64jd2k9i0cc3rc14-eta-build-0.0.0.drv' failed with exit code 1
error: build of '/nix/store/xrg25dw736rz3sns88g6p0fqwjwvizzv-eta-0.7.2.1.drv', '/nix/store/xvvandq9rzf0z3gw64jd2k9i0cc3rc14-eta-build-0
.0.0.drv' failed
brandon@brandon-750-170se-DevContainer:~/workspace/ghcvm
```

**Comment by puffnfresh** [comment]
Wow that's really useful @bbarker, I'll take a look.

**Comment by puffnfresh** [comment]
@bbarker a big part of the problem is ghc7103 is not longer supported in nixpkgs, even though it's still there. Many packages are missing the dependencies they need. I managed to work around most of them but got stuck on bin-package-db.

@rahulmutt there's no chance of Eta compiling under ghc802 or newer, right?

**Comment by rahulmutt** [comment]
@puffnfresh There is a chance. We face long compile times when hacking on the compiler and are concerned about the compile-time regressions in GHC 8. It doesn't hurt to experiment though and bump up the stack resolver to the most recent version and see what happens.

GHC 8 has a better optimization pipeline so that *may* mean that the Eta compiler itself will get a nice perf boost, but we'll have to see.

**Comment by jwiegley** [comment]
Has there been any progress on this? I'm also having difficulties getting it to build with current nixpkgs.

**Comment by puffnfresh** [comment]
Upgrading to GHC 8.2 is fairly involved. I've instead rewritten the Nix expression to use nixpkgs 17.09 so that we get GHC 7.10 instead:

#781 

This gets us really close to something working again.

**Comment by jwiegley** [comment]
@puffnfresh Very close! Though now it fails compiling Eta itself:

```
[173 of 304] Compiling Eta.Main.HscTypes ( compiler/Eta/Main/HscTypes.hs, dist/build/Eta/Main/HscTypes.o )

compiler/Eta/Main/HscTypes.hs:2966:56:
    Not in scope: ‘classFileName’
```

**Comment by jwiegley** [comment]
Never mind, didn't update submodules.

**Comment by puffnfresh** [comment]
Fixed by #781
